src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,137 @@
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    MyLinkedList() {
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        indexCheck(index);
+        Node<T> newNode = new Node<>(null,value, head);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else {
+            Node<T> prev = getNode(index - 1);
+            newNode.next = prev.next;
+            prev.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        Node<T> current = getNode(index);
+        T temp = current.item;
+        current.item = value;
+        return temp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        T temp;
+        if (index == 0) {
+            temp = head.item;
+            head = head.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else {
+            Node<T> prev = getNode(index - 1);
+            temp = prev.next.item;
+            prev.next = prev.next.next;
+            if (index == size - 1) {
+                tail = prev;
+            }
+        }
+        size--;
+        return temp;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> temp = head;
+        for (int i = 0; i < size; i++) {
+            if (object == temp.item || object != null && object.equals(temp.item)) {
+                remove(i);
+                return true;
+            }
+            temp = temp.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        indexCheck(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private void indexCheck(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Incorrect index" + index);
```

Comment:
you can add size to message : `  throw new IndexOutOfBoundsException("Incorrect index " + index + " when index is" + size);` but it isn't necessarily

Reply:
it's probably dosn't help with "index = -7"

Reply:
Add more info to the exception message, because it doesn't contain enough info, why is the index incorrect? Also, don't forget about spaces. Now you have message "Incorrect index-7"